### PR TITLE
[CMSIS-NN] Convert CMSIS-NN to use Target Hooks

### PR DESCRIPTION
**Pull request by Mousius**
This migrates CMSIS-NN to use Target Hooks instead of fully BYOC, which
means it will now go through any central passes the Driver API.

Found a few things whilst doing this:
* Forgot to mutate PrimFunc arguments in LowerTE which meant functions weren't getting lowered passed the first function in test_networks
* Target `cmsis-nn` needs to match external code generator `cmsis-nn` to connect the Target with the external code generator
* Partition Graph needed to sanitise compiler names to generate them properly in C

### CODE REVIEW

**Review by ashutosh-arm — COMMENTED** [review]
Thanks @Mousius. LGTM. Minor comments in there.

File: src/relay/backend/contrib/cmsisnn/tir_to_runtime.cc [hunk 1]
```diff
@@ -25,114 +25,50 @@
...
 namespace tvm {
-namespace codegen {
-
 using namespace tir;
+namespace relay {
+namespace contrib {
+namespace cmsisnn {
 
-class CodeGenCMSISNN : public CodeGenC {
+class CodeGenCMSISNN : public codegen::CodeGenCHost {
```
[the remark this answers is not in the document]
Comment (Mousius) [reply src/relay/backend/contrib/cmsisnn/tir_to_runtime.cc:36]: In other `Target`s they get PrimFunc's which are essentially:
```
{ [Host Module] -> [Device Lib Module] } -> [Device]
```
So conceptually it makes sense to do:
```
{ [Host Module] } -> [Lib]
```
For CMSIS-NN and other library targets, as TVM is hosting the library - I think, would be interested in @manupa-arm's view here.

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc [hunk 2]
```diff
@@ -33,29 +34,46 @@ namespace relay {
 namespace contrib {
 namespace cmsisnn {
 
-class RelayToTIRVisitor : public MixedModeVisitor {
+class RelayToTIRVisitor : public MixedModeMutator {
```
[the remark this answers is not in the document]
Comment (Mousius) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:37]: I'll move this to a follow up :smile_cat:

**Review by manupa-arm — APPROVED** [review]
LGTM modulo the existing comments.

**Comment by Mousius** [comment]
@ashutosh-arm could you take a look and explicitly approve if it's ok to move the name change to a follow up :smile_cat:

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:37]: Sorry, I didn't mean it as a blocking comment. A follow up is fine, should've marked as nit.

File: src/relay/backend/contrib/cmsisnn/tir_to_runtime.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (manupa-arm) [reply src/relay/backend/contrib/cmsisnn/tir_to_runtime.cc:36]: This seems fine to me.

**Comment by manupa-arm** [comment]
Thanks @Mousius @ashutosh-arm !

**Pull request merged**